### ISSUE TITLE
THREADING ERROR false positive when using vkWaitForPresentKHR

### ISSUE THREAD

**Issue by filnet** [KhronosGroup/Vulkan-ValidationLayers#3615]
**Describe the Issue**
Calling `vkWaitForPresentKHR` from a separate thread generates a THREADING ERROR:

```
 [Error][Validation]"Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0x94f99f0000008f6a, type = VK_OBJECT_TYPE_SWAPCHAIN_KHR; | MessageID = 0x141cb623 |THREADING ERROR : vkAcquireNextImageKHR(): object of type VkSwapchainKHR is simultaneously used in thread 0x41e4 and thread 0x2ccc"
```

According to [vkWaitForPresentKHR](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkWaitForPresentKHR.html), there is an exception for this function :

> As an exception to the normal rules for objects which are externally synchronized, the swapchain passed to vkWaitForPresentKHR may be simultaneously used by other threads in calls to functions other than vkDestroySwapchainKHR. Access to the swapchain data associated with this extension must be atomic within the implementation.

**Environment:**
 - OS: Windows 10
 - GPU: RTX 2060
 - SDK or header version if building from repo: 1.2.198.0
 - Options enabled (synchronization, best practices, etc.): synchronization, best practices
 -

**Comment by filnet** [comment]
I don't mind giving a shot at addressing this issue if I am given some guidance (i.e. general approach + entry points in code).

**Comment by artem-lunarg** [comment]
Raising this in priority and can check the next week.
Some details if you are interesting to check implementation. Threading validation is partially generated code (by the python script).

For example in `layers\vulkan\generated\thread_safety.cpp` we have this generated code:
```
void ThreadSafety::PreCallRecordWaitForPresentKHR(VkDevice device, VkSwapchainKHR swapchain, uint64_t presentId, uint64_t timeout,
                                                  const RecordObject& record_obj) {
    StartReadObjectParentInstance(device, record_obj.location);
    StartWriteObject(swapchain, record_obj.location);
    // Host access to swapchain must be externally synchronized
}

void ThreadSafety::PostCallRecordWaitForPresentKHR(VkDevice device, VkSwapchainKHR swapchain, uint64_t presentId, uint64_t timeout,
                                                   const RecordObject& record_obj) {
    FinishReadObjectParentInstance(device, record_obj.location);
    FinishWriteObject(swapchain, record_obj.location);
    // Host access to swapchain must be externally synchronized
}
```

And the python script that generates is here:
`scripts\generators\thread_safety_generator.py`

StartWrite/FinishWrite object mark regions when the object is in use by the current API function call. If two such Start/Finish sections from the different threads overlap this generated a threading error.

And thanks for the reminder.

**Comment by filnet** [comment]
Found more special cases:

**vkCreate*Pipelines functions**

> If pipelineCache was created with VK_PIPELINE_CACHE_CREATE_EXTERNALLY_SYNCHRONIZED_BIT, host access to pipelineCache must be [externally synchronized](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#fundamentals-threadingbehavior).

Examples:
* https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCreateComputePipelines-pipelineCache-02873
* https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCreateGraphicsPipelines-pipelineCache-02876

**vkCreatePipelineCache**

> If flags of pCreateInfo includes VK_PIPELINE_CACHE_CREATE_EXTERNALLY_SYNCHRONIZED_BIT, all commands that modify the returned pipeline cache object must be [externally synchronized](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#fundamentals-threadingbehavior).

**VkCommandBuffer**

Most commands that take a command buffer as argument have this additional requirement:

> Host access to the VkCommandPool that commandBuffer was allocated from must be externally synchronized

**Other**

* [Deferred Host Operations](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#deferred-host-operations-requesting)
* https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkUpdateDescriptorSets-pDescriptorWrites-06993
* [vkDestroyDebugUtilsMessengerEXT](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkDestroyDebugUtilsMessengerEXT.html)
* [vkDebugMarkerSetObjectTagEXT](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html
* /vkDebugMarkerSetObjectTagEXT.html)

This is list is probably not exhaustive.

**Comment by artem-lunarg** [comment]
I'll have a PR soon with a fix for vkWaitForPresentKHR which will close this ticket.
The use cases you found we'll add to a todo list. Thanks for looking for this case!

**Comment by filnet** [comment]
I humbly suggest to rename the generator to `host_synchronization_generator.py`

**Comment by artem-lunarg** [comment]
Thread-safety it's the established name for this functionality of this project. Rename won't happen.

**Comment by artem-lunarg** [comment]
Added new item to Validation Layer Enhancements project:
https://github.com/orgs/KhronosGroup/projects/22?pane=issue&itemId=44076637
